### ISSUE TITLE
Workflow issue with ExternalFile. 

### ISSUE THREAD

**Issue by MingboPeng** [NREL/OpenStudio#3608]
Hi All,

I am currently trying to use ExternalFile.getExternalFile() right after Model is constructed, and there is an issue related to this workflow. 

Here is a short c# code example:
```
Model m = new Model();
var optionalFile = ExternalFile.getExternalFile(m, csv);
.....
m.save(OSMPATH,true);
```

In ExternalFile current constructor, it only takes the filename string, instead of absolute filepath, or Path object. It uses `workflowJSON` to complete the filepath, and this will always cannot find the file or copy the file, because there are no OSM file and workflowJSON exists yet. 

In other words, using this ExternalFile.getExternalFile() to construct ExternalFile won't work with the Model object initiated in the same code.

https://github.com/NREL/OpenStudio/blob/e8fd0b1b3df91d61baa4c94d4b8b41e3d4e164cd/openstudiocore/src/model/ExternalFile.cpp#L202-L219

Please let me know if this doesn't make sense.

Thanks in advance,
Mingbo

**Comment by macumber** [comment]
Hey @MingboPeng thanks for the report.  I'm not sure it's valid though.  The following Ruby code works for me.  You can put your file in one of the WorkflowJSON file paths and find it by name.  Otherwise, you can also pass an absolute path, [findFile](https://github.com/NREL/OpenStudio/blob/e077c7edfa6c94bb37aeb0e31f40b69ed2740223/openstudiocore/src/utilities/filetypes/WorkflowJSON.cpp#L495) will accept an absolute path.

```
require 'openstudio'

m = OpenStudio::Model::Model.new

# csv file will be copied to first file path, you could put your file in any of these paths and find it by name
puts "[#{m.workflowJSON.filePaths.join(', ')}]" 

# alternatively you can use an absolute path, the file will be copied
f = OpenStudio::Model::ExternalFile.getExternalFile(m, 'D:/test/schedule_file.csv')

m.save('./schedule_file.osm',true)
```

**Comment by MingboPeng** [comment]
Hi @macumber , thanks for testing this issue. I am using the absolute path in getExternalFile(). But when it saved to osm file, the ExternalFile's file name only shows the file name instead of absolute file path. 
```
OS:External:File,
  {2d56715c-9e28-4986-ab79-fe74a689da60}, !- Handle
  External File 1,                        !- Name
  SCHtest.csv;                            !- File Name
```

The result of running the simulation is it failed to find the file.
![image](https://user-images.githubusercontent.com/9031066/62176657-59442e80-b310-11e9-836f-dab6496bea95.png)

I noticed the file path for EPW file, it uses the full absolute file path when it just saved in osm file, and changed to relative path `file:files/USA_NY_New.York-Central.Park.725033_TMY3.epw` after run. 
Is it possible to have the same as EPW for ExternalFile?

![image](https://user-images.githubusercontent.com/9031066/62176818-bfc94c80-b310-11e9-84b4-68baa9cd5337.png)

![image](https://user-images.githubusercontent.com/9031066/62176796-b63fe480-b310-11e9-9ec8-cb69b769a12d.png)

Thanks again.
Mingbo

**Comment by macumber** [comment]
I think I see what is happening here.  Can you confirm if this is the workflow you are using?

1. Generate an OSM with external file in some script or code
1. Open that OSM in the OpenStudio Application
1. Hit run button in the OpenStudio Application

**Comment by MingboPeng** [comment]
Correct. 
I think based on the OSM file only, without WorkflowJSON file, it cannot locate the absolute file path.
Saving the absolute file path like EPW file might just solve the problem.
Thanks.

**Comment by macumber** [comment]
I would recommend creating the companion directory when you save your model, then it will be detected when you load the model from the CLI or in the app:

```
require 'openstudio'

osm_path = './schedule_file.osm'
companion_dir = OpenStudio::getCompanionFolder(osm_path)
osw_path = File.join(companion_dir.to_s, 'workflow.osw')

m = OpenStudio::Model::Model.new
m.workflowJSON.saveAs(osw_path)

# csv file will be copied to first file path, you could put your file in any of these paths and find it by name
puts "[#{m.workflowJSON.filePaths.join(', ')}]" 

# alternatively you can use an absolute path, the file will be copied
f = OpenStudio::Model::ExternalFile.getExternalFile(m, 'D:/test/schedule_file.csv')

m.save('./schedule_file.osm',true)

m2 = OpenStudio::Model::Model.load(osm_path).get
m2.getExternalFiles.each do |ef|
  puts ef.filePath
end
```

**Comment by MingboPeng** [comment]
Thanks @macumber ,
I just tested and run in app, it works.
But with CLI, it seems still a problem. I checked the folder, the csv file is inside, it seems the path in following error missing the file extension ".csv".
![image](https://user-images.githubusercontent.com/9031066/62179103-539f1680-b319-11e9-967b-894917531292.png)

**Comment by macumber** [comment]
I just updated the code above, can you try the above (adding weather file, etc as needed) and by running the OSW file like:

`/openstudio run -w schedule_file/workflow.osw`

**Comment by MingboPeng** [comment]
Thanks @macumber , I think it works so far. There is also a bug in Honeybee that removes ".csv" extension https://github.com/mostaphaRoudsari/honeybee/issues/727.
